sim-lib/src/cln.rs

```diff
@@ -101,10 +107,47 @@ impl LightningNode for ClnNode {
 
     async fn track_payment(
         &mut self,
-        _hash: PaymentHash,
-        _shutdown: Listener,
+        hash: PaymentHash,
+        shutdown: Listener,
     ) -> Result<PaymentResult, LightningError> {
-        unimplemented!()
+        loop {
+            tokio::select! {
+                biased;
+                _ = shutdown.clone() => {
+                    break Err(LightningError::TrackPaymentError("Shutdown before tracking results".to_string()));
+                },
+                _ = time::sleep(Duration::from_secs(1)) => {
+                    let ListpaysResponse { pays } = self
+                        .client
+                        .list_pays(ListpaysRequest {
+                            payment_hash: Some(hash.0.to_vec()),
+                            ..Default::default()
+                        })
+                        .await
+                        .map_err(|err| LightningError::TrackPaymentError(err.to_string()))?
+                        .into_inner();
+
+                    if let Some(pay) = pays.first() {
+                        let payment_status = ListpaysPaysStatus::from_i32(pay.status)
+                            .ok_or(LightningError::TrackPaymentError("Invalid payment status".to_string()))?;
+
+                        let payment_outcome = match payment_status {
+                            ListpaysPaysStatus::Pending => continue,
+                            ListpaysPaysStatus::Complete => PaymentOutcome::Success,
+                            // Task: https://github.com/bitcoin-dev-project/sim-ln/issues/26#issuecomment-1691780018
+                            ListpaysPaysStatus::Failed => PaymentOutcome::UnexpectedError,
+                        };
+                        let htlc_count = pay.number_of_parts.unwrap_or(1).try_into().map_err(|_| LightningError::TrackPaymentError("Invalid number of parts".to_string()))?;
+                        break Ok(PaymentResult {
```

Comment:
Is that break returning? Please don't. This is literally the return of the method.

Reply:
return works, but imo `break val` here better communicates the exit condition of the outer `loop`. its really easy to miss the loop.

Reply:
Sure, but there is nothing outside the outer loop, nor are you using the value returned by the loop to anything but returning from the method, hence what I think returning from there makes more sense

Reply:
makes sense